[systemd] collect taskCurrent, tasksMax per systemd unit

auhlig opened the pull request:
Adds the current and max tasks for a systemd unit. LGTY @SuperQ?

Signed-off-by: Arno Uhlig <arno.uhlig@sap.com>

auhlig commented on collector/systemd_linux.go:
```diff
@@ -258,6 +295,29 @@ func (c *systemdCollector) getAllUnits() ([]unit, error) {
+			} else {
+				val := tasksMaxCount.Value.Value().(uint64)
+				// don't set if tasksMax if dbus reports MaxUint64
+				if val != math.MaxUint64 {
```
not going to report a metric if there's no limit configured

auhlig commented on collector/systemd_linux.go:
```diff
@@ -258,6 +295,29 @@ func (c *systemdCollector) getAllUnits() ([]unit, error) {
+				val := tasksCurrentCount.Value.Value().(uint64)
+				// don't set if tasksCurrent if dbus reports MaxUint64
+				if val != math.MaxUint64 {
+					unit.tasksCurrent = &val
```
don't set and don't report a metric if the TasksCurrent is MaxUint64 as seen here:
```
busctl introspect org.freedesktop.systemd1 /org/freedesktop/systemd1/unit/systemd_2drandom_2dseed_2eservice | grep Tasks
.TasksAccounting                    property  b              true                                     -
.TasksCurrent                       property  t              18446744073709551615                     -
.TasksMax                           property  t              81919
```
not sure that's the right way to handle this. maybe we should report 0 or -1 ?

auhlig replied:
That's what I was asking myself when I saw it. MaxUint64 could be explained as infinite for TasksMax, but I cannot see a rationale for TasksCurrent = MaxUint64 :man_shrugging: 
I just would report it in this case

SuperQ replied:
Yes, let's just report `TasksCurrent` as-is. We generally want to avoid transforming metrics from another system.

auhlig commented:
I guess codespell needs some training: `./collector/nfs_linux.go:141: Uint  ==> Unit` ;)

SuperQ commented:
This is starting to leak out of systemd-specifics and more into container metrics. I think this would be better solved by using [cAdvisor](https://github.com/google/cadvisor).

SuperQ commented:
FYI, a rebase against master will fix the codespell problem.

auhlig commented:
Rebased.
Thanks for taking the time and sorry for the delay @SuperQ.
While I agree that this should find its way to cAdvisor to collect the number in a k8s env, I'd also think this could be useful in the node exporter on more platforms. 
Was quite hard to find units leaking threads and there maximum without this.

auhlig commented:
I also had a look after you mentioned it. cAdvisor is completely missing metrics for this

SuperQ commented:
Please add an `[ENHANCEMENT]` entry to CHANGELOG.md

SuperQ commented:
It would also be good to add tests for this in `systemd_linux_test.go`.

auhlig commented:
rebased and updated changelog. 
currently there's only a very high level test for systemd [here](https://github.com/prometheus/node_exporter/blob/master/collector/systemd_linux_test.go#L90-L108). i'm not sure how to properly test this further with reasonable effort. maybe reading the metrics from the `sink` and comparing to a fixture?

SuperQ commented on collector/systemd_linux.go:
```diff
@@ -258,6 +295,29 @@ func (c *systemdCollector) getAllUnits() ([]unit, error) {
+				log.Debugf("couldn't get unit '%s' TasksCurrent: %s", unit.Name, err)
+			} else {
+				val := tasksCurrentCount.Value.Value().(uint64)
+				// don't set if tasksCurrent if dbus reports MaxUint64
```
Please make sure all comments are complete sentences. IE. `Don't ... MaxUint64.`

SuperQ commented on CHANGELOG.md:
```diff
@@ -43,6 +43,7 @@ Darwin meminfo metrics have been renamed to match Prometheus conventions. #1060
 * [ENHANCEMENT] Handle stuck NFS mounts #997
 * [ENHANCEMENT] infiniband: Handle iWARP RDMA modules N/A #974
 * [ENHANCEMENT] Update diskstats for linux kernel 4.19 #1109
+* [ENHANCEMENT] Collect TasksCurrent, TasksMax per systemd unit
```
This should in include the PR number, `#1098`.

auhlig replied:
Oje. Missed this

auhlig commented:
Thanks @SuperQ @discordianfish ❤️

edolce62 commented:
Hi guys,

I still do not see the taskCurrent, tasksMax per systemd unit in prometheus, I don't see the debug messages below:

log.Debugf("couldn't get unit '%s' TasksMax: %s", unit.Name, err)
log.Debugf("couldn't get unit '%s' TasksCurrent: %s", unit.Name, err)

so I guess the collection does not fail.

Looking at the source code I read:

		// Don't set if tasksMax if dbus reports MaxUint64.
		if val != math.MaxUint64 {
			unit.tasksMax = &val

can be this the issue?

Any help will be much appreciated, thanks !

Eric

SuperQ commented:
@edolce62 Please do not post help questions to completed pull requests.

For usage questions/help, please use our [community](https://prometheus.io/community/). On the mailing list, more people are available to potentially respond to your question, and the whole community can benefit from the answers provided.

edolce62 commented:
@SuperQ Sorry for that, I posted there.
